play-json/src/main/scala/com/lightbend/lagom/scaladsl/playjson/PlayJsonSerializer.scala

```diff
@@ -66,16 +67,17 @@ private[lagom] final class PlayJsonSerializer(val system: ExtendedActorSystem, r
   override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = {
     val startTime = if (isDebugEnabled) System.nanoTime else 0L
 
-    val i = manifest.lastIndexOf('#')
-    val fromVersion = if (i == -1) 1 else manifest.substring(i + 1).toInt
-    val manifestClassName = if (i == -1) manifest else manifest.substring(0, i)
+    val (fromVersion: Int, manifestClassName: String) = parseManifest(manifest)
 
     val migration = migrations.get(manifestClassName)
 
     val migratedManifest = migration match {
-      case Some(migration) if migration.currentVersion > fromVersion =>
+      case Some(migration) if (migration.currentVersion > fromVersion) =>
         migration.transformClassName(fromVersion, manifestClassName)
-      case None => manifestClassName
+      case Some(migration) if (migration.currentVersion < fromVersion) =>
+        throw new IllegalStateException(s"Migration version ${migration.currentVersion} is " +
+          s"behind version $fromVersion of deserialized type [$manifestClassName]")
```

Comment:
Failing here might be the right thing to do - but it does potentially make things like canary deploys dangerous, since events will be written out in the new version, but then if the canary deploy fails, those entities are now corrupted... I wonder if a more pragmatic solution is to try to deserialize with the old version (which in many cases, may succeed just fine, eg if a migration simply adds a new required field, there's no reason why that wouldn't work - of course, there are other situations where it could be dangerous).  Maybe the problem shouldn't be solved here, perhaps a backwards migrations feature should be added, where for a time, events are back migrated before being persisted, until all nodes are upgraded, and then when that deploy is successful, all nodes can be upgraded.